topology/probes/libvirt/libvirt.go

```diff
@@ -0,0 +1,267 @@
+	cidDevAdded                int                   // second monitor on devices added to domains
+	uri                        string                // uri of the libvirt connection
+}
+
+// Mac describes the XML coding of the guest mac address of an interface in
+// libvirt
+type Mac struct {
+	Address string `xml:"address,attr"`
+}
+
+// Target describes the XML coding of the target of an interface in libvirt
+// The target is the Linux name on the host.
+type Target struct {
+	Device string `xml:"dev,attr"`
+}
+
+// Alias defines the XML coding of the libvirt name of the interface
+type Alias struct {
+	Name string `xml:"name,attr"`
+}
+
+// Address describes the XML coding of the pci addres of an interface in libvirt
+type Address struct {
+	Type     string `xml:"type,attr"`
+	Domain   string `xml:"domain,attr"`
+	Bus      string `xml:"bus,attr"`
+	Slot     string `xml:"slot,attr"`
+	Function string `xml:"function,attr"`
+}
+
+// Interface is XML coding of an interface in libvirt
+type Interface struct {
+	Mac     Mac     `xml:"mac"`
+	Target  Target  `xml:"target"`
+	Address Address `xml:"address"`
+	Alias   Alias   `xml:"alias"`
+	Host    string  `xml:"-"`
+}
+
+// Domain is the subset of XML coding of a domain in libvirt
+type Domain struct {
+	Interfaces []Interface `xml:"devices>interface"`
+}
+
+// queryDomain uses libvirt to get information on the interfaces of a domain.
+// domain is the libvirt domain queried and bag is an accumulator for results
+func (probe *Probe) queryDomain(
+	domain *libvirtgo.Domain, // domain to query
+	bag *[]*Interface, // to accumulate results
+	constraint string, // to restrict the search to a single interface (by alias)
+) {
+	rawXML, err := domain.GetXMLDesc(0)
+	if err != nil {
+		logging.GetLogger().Error("Cannot get XMLDesc: %s", err)
+		return
+	}
+	d := Domain{}
+	if err = xml.Unmarshal([]byte(rawXML), &d); err != nil {
+		logging.GetLogger().Error("XML parsing error: %s", err)
+		return
+	}
+	host, _ := domain.GetName()
+	for i := range d.Interfaces {
+		itf := &d.Interfaces[i]
+		if constraint == "" || constraint == itf.Alias.Name {
+			itf.Host = host
+			*bag = append(*bag, itf)
+		}
+	}
+}
+
+// registerInterfaces puts the information collected in the graph
+// interfaces is an array of collected information.
+func (probe *Probe) registerInterfaces(interfaces []*Interface) {
+	probe.graph.Lock()
+	defer probe.graph.Unlock()
+	for _, itf := range interfaces {
+		name := itf.Target.Device
+		if name == "" {
+			continue
+		}
+		probe.interfaceMap[name] = itf
+		logging.GetLogger().Debugf(
+			"Libvirt interface %s on %s", name, itf.Host)
+		node := probe.graph.LookupFirstNode(
+			graph.Metadata{"Type": "tun", "Name": name})
+		if node != nil {
+			probe.enrichNode(node, itf)
+		}
+	}
+}
+
+// enrichNode adds the libvirt interface information to a node of the graph
+func (probe *Probe) enrichNode(node *graph.Node, itf *Interface) {
```

Comment:
I do love this extended metadata information. We can move it here: https://github.com/skydive-project/skydive/pull/1424